### PR TITLE
Fix login logout when engine is down and credential helper is in use

### PR DESCRIPTION
**Pull request by djs55**
**- What I did**

Fixed `docker login` and `docker logout` when the engine is not running and when using a credential helper.

**- How I did it**

The logic in `ElectAuthServer` accidentally combined together checking the error from the engine with the state of the debug flag when deciding which value to return.

**- How to verify it**

I used Docker Desktop on Mac with the `docker-credential-osxkeychain` enabled (which is the default). I shutdown the app. Without this patch I ran:

```
$ docker logout
Not logged in to
```
-- note the missing server URL

```
$ docker --debug logout
Warning: failed to get default registry endpoint from daemon (Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?). Using system default: https://index.docker.io/v1/
Not logged in to https://index.docker.io/v1/
```    
-- note the server URL is present

This shows that the `--debug` flag is changing the actual behaviour of the program.

I ran `docker login` and observed an error from the credential helper
```
Error saving credentials: error storing credentials - err: no credentials server URL, out: `no credentials server URL`
```

When I rebuilt the CLI with this PR `docker login` and `docker logout` operated as expected.

I separated this PR into 2 patches:
1. the actual bugfix
2. a small refactoring which I think makes `ElectAuthServer` easier to read but which is optional

**- Description for the changelog**
Fix `docker login` and `docker logout` when the engine is not running and a credential helper is in use.

**- A picture of a cute animal (not mandatory but encouraged)**

39df0ec8dab014008eb217865bc4635b

### CODE REVIEW

File: cli/command/registry.go [hunk 1]
```diff
@@ -28,16 +28,22 @@ func ElectAuthServer(ctx context.Context, cli Cli) string {
...
 	// the default registry URL might be Windows specific.
-	serverAddress := registry.IndexServer
-	if info, err := cli.Client().Info(ctx); err != nil && debug.IsEnabled() {
-		// Only report the warning if we're in debug mode to prevent nagging during engine initialization workflows
-		fmt.Fprintf(cli.Err(), "Warning: failed to get default registry endpoint from daemon (%v). Using system default: %s\n", err, serverAddress)
-	} else if info.IndexServerAddress == "" && debug.IsEnabled() {
-		fmt.Fprintf(cli.Err(), "Warning: Empty registry endpoint from daemon. Using system default: %s\n", serverAddress)
-	} else {
-		serverAddress = info.IndexServerAddress
+	info, err := cli.Client().Info(ctx)
```
Comment (thaJeztah) [inline cli/command/registry.go]: Should we use `/_ping` for this? (`/info` is a bit heavy _just_ to know if the engine is up and running)

File: cli/command/registry.go [hunk 2]
```diff
@@ -28,16 +28,22 @@ func ElectAuthServer(ctx context.Context, cli Cli) string {
...
+	if err != nil {
+		// Daemon is not responding so use system default.
+		if debug.IsEnabled() {
+			// Only report the warning if we're in debug mode to prevent nagging during engine initialization workflows
+			fmt.Fprintf(cli.Err(), "Warning: failed to get default registry endpoint from daemon (%v). Using system default: %s\n", err, registry.IndexServer)
+		}
+		return registry.IndexServer
+	}
+	if info.IndexServerAddress == "" {
```
Comment (thaJeztah) [inline cli/command/registry.go]: Oh, never mind, I see we need this information 🤔

Comment (thaJeztah) [reply cli/command/registry.go]: Actually looks like it's hard-coded, and never updated anywhere 🤔 https://github.com/moby/moby/blob/fc01c2b481097a6057bec3cd1ab2d7b4488c50c4/registry/config.go#L29-L48

File: cli/command/registry.go [hunk 3]
```diff
@@ -28,16 +28,22 @@ func ElectAuthServer(ctx context.Context, cli Cli) string {
...
+		if debug.IsEnabled() {
```
Comment (silvin-lubecki) [inline cli/command/registry.go]: We definitely need a real log policy here: we should write something like `log.Warning(...)` and let the debug flag change only the output level, not spreading it everywhere. But of course that's another issue.

File: cli/command/registry.go [hunk 4]
```diff
@@ -28,16 +28,22 @@ func ElectAuthServer(ctx context.Context, cli Cli) string {
 	// used. This is essential in cross-platforms environment, where for
 	// example a Linux client might be interacting with a Windows daemon, hence
```
Comment (thaJeztah) [inline cli/command/registry.go]: Looks like this was added in https://github.com/moby/moby/pull/19891 at a time where a temporary/special Windows registry was set up: https://github.com/moby/moby/issues/18019

Actually thinking if that handling is now just dead code and no longer needed

**Comment by djs55** [comment]
Thanks for the very quick response (and merge)!

**Pull request merged**